keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt

```diff
@@ -1042,18 +1044,27 @@ class SqlDeliveryConfigRepository(
         select(DELIVERY_CONFIG.UID, DELIVERY_CONFIG.NAME)
           .from(DELIVERY_CONFIG, DELIVERY_CONFIG_LAST_CHECKED)
           .where(DELIVERY_CONFIG.UID.eq(DELIVERY_CONFIG_LAST_CHECKED.DELIVERY_CONFIG_UID))
+          .and(
+            DELIVERY_CONFIG_LAST_CHECKED.LOCKED_BY.isNull
+              .or(DELIVERY_CONFIG_LAST_CHECKED.LOCKED_AT.lessOrEqual(cutoff))
+          )
           .and(DELIVERY_CONFIG_LAST_CHECKED.AT.lessOrEqual(cutoff))
+          .andNotExists(
+            selectOne()
+              .from(PAUSED)
+              .where(PAUSED.NAME.eq(DELIVERY_CONFIG.APPLICATION))
+              .and(PAUSED.SCOPE.eq(APPLICATION.name))
+          )
           .orderBy(DELIVERY_CONFIG_LAST_CHECKED.AT)
           .limit(limit)
           .forUpdate()
           .fetch()
           .also {
             it.forEach { (uid, _) ->
-              insertInto(DELIVERY_CONFIG_LAST_CHECKED)
-                .set(DELIVERY_CONFIG_LAST_CHECKED.DELIVERY_CONFIG_UID, uid)
-                .set(DELIVERY_CONFIG_LAST_CHECKED.AT, now.toTimestamp())
-                .onDuplicateKeyUpdate()
-                .set(DELIVERY_CONFIG_LAST_CHECKED.AT, now.toTimestamp())
+              update(DELIVERY_CONFIG_LAST_CHECKED)
+                .set(DELIVERY_CONFIG_LAST_CHECKED.LOCKED_BY, InetAddress.getLocalHost().hostName)
+                .set(DELIVERY_CONFIG_LAST_CHECKED.LOCKED_AT, now.toTimestamp())
+                .where(DELIVERY_CONFIG_LAST_CHECKED.DELIVERY_CONFIG_UID.eq(uid))
```

Comment:
Instead of updating the checked at timestamp right away, mark the delivery config as locked (using local host name is just convenient for debugging, the value is arbitrary). Also added a `locked_at` time so we can ignore stale locks.

Reply:
As an aside, I really like the term [leases](http://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html) for the kinds of lock implemented here, that expires after a certain amount of time.

Reply:
Yeah, that's good. I'll change it